### PR TITLE
driver/docker: enable setting hard/soft memory limits

### LINKED ISSUE #2093: [question] Launch docker container with hard memory limit higher than specified in task resources
We are using Nomad do deploy microservices wrapped in docker containers.

But memory consumption of microservices is non-uniform. 
For example, a microservice can consume on average, say, 50mb. But there is a heavy endpoint which is rarely called and consumes, say, 100mb.

We have to specify memory limit based on worst case. Most of the time memory is not fully utilized and we have to pay for more aws instances than we actually need.

Is there any way to tell Nomad "when planning microservice allocation assume 50mb usage, but actually launch container with 100mb limit / without memory limi "?

### PR DESCRIPTION
**Pull request by shoenig**
Fixes #2093

Enable configuring `memory_hard_limit` in the docker config stanza for tasks.
If set, this field will be passed to the container runtime as `--memory`, and
the `memory` configuration from the task resource configuration will be passed
as `--memory_reservation`, creating hard and soft memory limits for tasks using
the docker task driver.

### CODE REVIEW

**Comment by shoenig** [comment]
Example)

```bash
$ cat example.nomad 
job "example" {
  datacenters = ["dc1"]

  group "cache" {
    task "redis" {
      driver = "docker"

      config {
        image = "redis:3.2"

	memory_hard_limit = 512 # new!

        port_map {
          db = 6379
        }
      }

      resources {
        cpu    = 500
        memory = 256

        network {
          mbits = 10
          port "db" {}
        }
      }
    }
  }
}
```

`memory` and `memory_reservation` are now hard / soft values
```
$ docker inspect 1f6 | jq .[].HostConfig.Memory
536870912
$ docker inspect 1f6 | jq .[].HostConfig.MemoryReservation
268435456
```

Environment variable remains on soft limit, as staying under that limit is the intent 
```bash
$ docker inspect 1f6 | jq .[].Config.Env | grep NOMAD_MEMORY_LIMIT
  "NOMAD_MEMORY_LIMIT=256",
```

File: CHANGELOG.md [hunk 1]
```diff
@@ -11,6 +11,7 @@ IMPROVEMENTS:
  * csi: Return better error messages [[GH-7984](https://github.com/hashicorp/nomad/issues/7984)] [[GH-8030](https://github.com/hashicorp/nomad/issues/8030)]
  * csi: Move volume claim releases out of evaluation workers [[GH-8021](https://github.com/hashicorp/nomad/issues/8021)]
  * csi: Added support for `VolumeContext` and `VolumeParameters` [[GH-7957](https://github.com/hashicorp/nomad/issues/7957)]
+ * driver/docker: Added support for memory_hard_limit configuration in docker task driver [[GH-2093](https://github.com/hashicorp/nomad/issues/2093)]
```
Comment (schmichael) [inline CHANGELOG.md:14]:
```suggestion
 * driver/docker: Added support for `memory_hard_limit` configuration in docker task driver [[GH-2093](https://github.com/hashicorp/nomad/issues/2093)]
```

File: website/pages/docs/drivers/docker.mdx [hunk 2]
```diff
@@ -183,6 +183,13 @@ The `docker` driver supports the following configuration in the job spec. Only
 - `mac_address` - (Optional) The MAC address for the container to use (e.g.
   "02:68:b3:29:da:98").
 
+- `memory_hard_limit` - (Optional) The maximum allowable amount of memory used
+  (megabytes) by the container. If set, the [`memory`](http://localhost:3000/docs/job-specification/resources#memory)
```
Comment (schmichael) [inline website/pages/docs/drivers/docker.mdx:187]:
```suggestion
  (megabytes) by the container. If set, the [`memory`](/docs/job-specification/resources#memory)
```

**Comment by shoenig** [comment]
From the [docker](https://docs.docker.com/config/containers/resource_constraints/#limit-a-containers-access-to-memory) and [cgroup](https://www.kernel.org/doc/Documentation/cgroup-v1/memory.txt) docs, it seems like #1 is a weak "yes".

> Allows you to specify a soft limit smaller than --memory which is activated when Docker detects contention or low memory on the host machine.

> When the system detects memory contention or low memory, control groups
are pushed back to their soft limits. If the soft limit of each control
group is very high, they are pushed back as much as possible to make
sure that one control group does not starve the others of memory.

File: drivers/docker/driver_test.go [hunk 3]
```diff
@@ -2681,3 +2681,19 @@ func TestDockerDriver_CreateContainerConfig_CPUHardLimit(t *testing.T) {
...
+		require.Equal(t, int64(256*1024*1024), memory)
+		require.Equal(t, int64(0), memoryReservation)
+	})
+
+	t.Run("driver hard limit is set", func(t *testing.T) {
+		memory, memoryReservation := new(Driver).memoryLimits(512, 256*1024*1024)
+		require.Equal(t, int64(512*1024*1024), memory)
+		require.Equal(t, int64(256*1024*1024), memoryReservation)
+	})
+}
```
Comment (schmichael) [inline drivers/docker/driver_test.go:2699]: I wonder if there's an integration test we could piggy back on to ensure that Docker is interpreting our limits as we expect. Similar to this: https://github.com/hashicorp/nomad/blob/v0.11.2/drivers/docker/driver_test.go#L1406-L1454

Not a big deal or blocker.

Comment (shoenig) [reply drivers/docker/driver_test.go:2722]: Ahh, I was looking for something like these and missed them. Added!

**Comment by analytically** [comment]
failed to create container: API error (400): Minimum memoryswap limit should be larger than memory limit, see usage

**Comment by analytically** [comment]
does this set memory.limit_in_bytes?

**Pull request merged**